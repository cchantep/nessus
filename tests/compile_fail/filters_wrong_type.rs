// This test verifies that using the wrong type with filter_builder will cause a compile error

use tnuctipun::{FieldWitnesses, MongoComparable};
use tnuctipun::filters::empty;
use serde::{Serialize, Deserialize};
#[derive(Debug, Clone, Serialize, Deserialize, FieldWitnesses, MongoComparable)]
struct Product {
    pub Price: f64,
}

fn main() {
    // This should fail to compile because we're trying to use a string for a f64 field
    let mut builder = empty::<Product>();
    builder.eq::<product_fields::Price, _>("wrong type".to_string());
}
