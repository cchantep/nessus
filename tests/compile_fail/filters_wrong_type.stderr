error[E0277]: the trait bound `Product: MongoComparable<f64, std::string::String>` is not satisfied
  --> tests/compile_fail/filters_wrong_type.rs:14:13
   |
14 |     builder.eq::<product_fields::Price, _>("wrong type".to_string());
   |             ^^ the trait `MongoComparable<f64, std::string::String>` is not implemented for `Product`
   |
   = help: the following other types implement trait `MongoComparable<A, B>`:
             `Product` implements `MongoComparable<f64, f32>`
             `Product` implements `MongoComparable<f64, f64>`
             `Product` implements `MongoComparable<f64, i16>`
             `Product` implements `MongoComparable<f64, i32>`
             `Product` implements `MongoComparable<f64, i64>`
note: required by a bound in `FilterBuilder::<T>::eq`
  --> src/filters.rs
   |
   |     pub fn eq<F, V>(&mut self, value: V) -> &mut Self
   |            -- required by a bound in this associated function
...
   |         T: HasField<F> + MongoComparable<T::Value, V>,
   |                          ^^^^^^^^^^^^^^^^^^^^^^^^^^^^ required by this bound in `FilterBuilder::<T>::eq`
